### DISCUSSION TITLE
Please add documentation on how to import and use the background plugin, and how to build custom controllers

### DISCUSSION THREAD

**Discussion by LiChoi** [react-page/react-page#923] [General] [answered]
I like what's offered here, but there are gaps in the documentation.

It wasn't obvious how I was supposed to import the background plugin. I had to look at the actual code on this github repo to figure out that "default" is a function that will create the plugin, hence I had to import it like this:

const backgroundPlugin = require('@react-page/plugins-background');

const cellPlugins = [backgroundPlugin.default()]; 

...which is a different pattern compared to how we import all the other built-in plugins.

Also, it would have been useful to see an example of how to build a custom component for the controls. I figured out how to do it, but would have been nice for the documentation to clarify we have to pass in onChange like so:

controls: {
    type: 'custom',
    Component: ({data, onChange}) => (<CustomPluginControls data={data} onChange={onChange} />)
  }

**Comment by macrozone** [comment, marked as answer]
hi @LiChoi ,

You indeed need to execute the background-plugin as it takes some configuration (as some other plugins like image), see this example:

https://github.com/react-page/react-page/blob/master/examples/plugins/cellPlugins.ts

we might consider to also accept functions that return a `CellPlugin` as plugins, so that it would automatically execute it.

> 
> Also, it would have been useful to see an example of how to build a custom component for the controls. I figured out how to do it, but would have been nice for the documentation to clarify we have to pass in onChange like so:
> 
> controls: {
> type: 'custom',
> Component: ({data, onChange}) => ()
> }

i strongly recommend to use `type: "autoform"` if your plugin takes data. It will automatically create a form for you and handles the onChange callback. You can also omit the controls property if your plugin does not take any data.
`type: "custom"` is only required, if you need to have very special controls that can't be done with `type: "autoform"`. If that is the case, you can call `onChange` in the component with new data for the plugin. The existing of this property is documented in the type definitions (with js docs). So if you are using typescript or a good IDE, you should see that this property exists.

I would also like to hear, why you think you need `type: "custom"`. I am asking because we want to make `type: "autoform"` even more powerful, so we like to hear other use cases

**Reply by LiChoi** [reply]
@macrozone Thank you for responding! 

My first preference was to use the autoform, however, there were certain input options that required more than just a string or dropdown input. An example is the color picker (like the one we see for background), unless that is an option I'm not aware of. Same thing for the custom slate rich text editor - it doesn't have the color picker input option that I'm aware of. 

The second issue I faced with the autoform was that I needed to bring in external data from redux store to populate the control options. I tried to do that by defining the plugin within the parent component that contains the Editor - that way, I could just pull in the data I need and pass it into the plugin definition. However, I couldn't connect redux to a functional component, and I couldn't convert the functional component (that contains the Editor) into a state-component because that would break 

const [value, setValue] = useState(null);

There probably is a solution, but I went with a custom controller as a quick and temporary workaround until I can figure out how to pull redux data into the autoform.

**Reply by macrozone** [reply]
you can actually define custom input components on the autoform. Here a real world example:


```

const plugin: CellPlugin<{
  employeeId: string;
}> = {
  Renderer,
  id: "employeeInfo",
  version: 1,

  title: "Employee Info",
  description: "Employee Info",
  controls: {
    type: "autoform",
    schema: {
      required: ["employeeId"],
      properties: {
        employeeId: {
          type: "string",
          uniforms: {
            component: EmployeeIdSelector,
          },
        },
      },
    },
  },
};

```

here EmployeeIdSelector could e.g.  read state from redux, apollo or similar. It can be any component. 

This Component needs to be wrapped with `import { connectField } from "uniforms";`:

```
import { connectField } from "uniforms";

const optionRenderer = (employee) =>
  `${employee.firstName} ${employee.lastName}`;

const EmployeeIdSelector = connectField(({ value, onChange }) => {

const options = .... // from apollo, redux or so

return <select value={value} onChange={e => onChange(e.target.value)}>{
  options.map(o => <option value={o.value} key={o.value}>{o.label}</option>
}
});

export default EmployeeIdSelector;
```

**Reply by macrozone** [reply]
you can of course also just mount a color picker. Maybe i'll add this to the examples

**Reply by LiChoi** [reply]
@macrozone Awesome! Thank you for pointing this out. I will refactor what I have to make use of this.

**Reply by macrozone** [reply]
@LiChoi I'll pushed a change that 

1.) adds a ready-to-use color picker field for autoforms
2.) updated the example to use it: https://github.com/react-page/react-page/commit/c6bde40a76f76e8ad248e9a318e3ba1e47769c85

it will be available in 1.4

**Reply by LiChoi** [reply]
Nice! Thank you!

**Reply by macrozone** [reply]
i also pushed some docs update: https://react-page.github.io/docs/#/custom-cell-plugins?id=autoform-controls
